### PR TITLE
support for ${buildType} and install

### PR DESCRIPTION
**Pull request by rtbo**
a few commits to solve ${buildType} (#19) and install (#20).
As `${buildType}` main purpose is to be used in `buildDirectory`, it is necessary to store `selectedBuildType` in a new persistent cache file on disk (.vscode/cmaketools.json), otherwise it is not possible to find the last build configuration from the cmake cache.
Alternative to this additional persistent cache would be to test for a cmake cache with each possible value of `${buildType}` and find the one with youngest mtime.

### CODE REVIEW

**Comment by vector-of-bool** [comment]
Sorry to be so slow to respond. Been busy. How thoroughly have you tested with these changes? There are a few things I'd tweak, but it looks pretty good.

File: src/cmake.ts [hunk 1]
```diff
@@ -317,24 +358,14 @@ export class CMakeTools {
...
-            self.selectedBuildType = 'None';
-        if (await self.isMultiConf())
-            self.selectedBuildType = 'Debug';
-        else
-            self.selectedBuildType = (await self.cache.get('CMAKE_BUILD_TYPE')).as<string>();
+        if (await self.cache.exists() && await self.isMultiConf())
+            if (self.selectedBuildType === 'None')
+                self.selectedBuildType = 'Debug';
+
         return self.selectedBuildType;
```
Comment (vector-of-bool) [inline src/cmake.ts]: The logic here seems to be very different from the old logic. It should probably check for the existence of ``.cmaketools.json``, as well as defaulting to the build type in ``CMakeCache.txt`` if the ext-cache is not found, and 'None' in the case that ``CMakeCache.txt`` is also missing.

File: src/cmake.ts [hunk 2]
```diff
@@ -736,9 +769,22 @@ export class CMakeTools {
...
+                .replace('${buildType}', this.selectedBuildType);
             settings_args.push("-D" + key + "=" + value);
         }
+        let prefix = self.config<string>("installPrefix");
+        if (prefix && prefix !== "") {
+            prefix = prefix
+                .replace('${workspaceRoot}', vscode.workspace.rootPath)
+                .replace('${buildType}', this.selectedBuildType);
+            if (prefix.includes(' '))
+                prefix = '"' + prefix + '"';
```
Comment (vector-of-bool) [inline src/cmake.ts]: Quoting strings is probably incorrect behavior. It should handle arguments with special characters just fine.

**Comment by rtbo** [comment]
Sorry for having let this stale. been busy too.
I have tested each new addition with GLFW source code on a linux laptop and also checked that I was not breaking anything (of course I can't tell if I've been exhaustive with this). Also tested on Windows.

File: src/cmake.ts (hunk 2 above)
Comment (rtbo) [inline src/cmake.ts]: Yes, that is correct. Special characters are handled by `child_process.spawn`.

**Comment by vector-of-bool** [comment]
In the interim, [this issue](https://github.com/vector-of-bool/vscode-cmake-tools/issues/27) appeared! It seems that ``child_process`` does not correctly handle arguments/commands with spaces. I'll need to do some more testing.

File: src/cmake.ts (hunk 1 above)
Comment (rtbo) [inline src/cmake.ts]: Yes I agree. In fact `.cmaketools.json` is already checked for in the constructor and defaults `selectedBuildType` to `None` (check the line where _`extCache` is constructed).
I pushed ca0cec417e5dd62a021434c66a7d6a09ae18e65e to load `selectedBuildType` from `CMakeCache.txt` if it is `None`

**Comment by vector-of-bool** [comment]
I'll merge this into a branch and play around with it to see if there are any behaviors that might need tweaking.

**Comment by rtbo** [comment]
great!

**Pull request merged**